### DISCUSSION TITLE
TasmoCompile v9.1.0 ESP32 - Solo1 won't built

### DISCUSSION THREAD

**Discussion by Mark-Muc** [benzino77/tasmocompiler#301] [General]
Hello .. I'd like to built a version for Solo1 (Shelly 2 pm). Using the precomiled version it works fine and i can flash the modul  ->working.
I'd like to use it as a shutter .. so i habe to compile my own version.
If i select  `Solo1` as option:
![image](https://user-images.githubusercontent.com/67410847/194744302-83c51ebc-bbf1-4744-8070-c14a6ad04a0e.png)

and then just to try to compile, i get the error:
![image](https://user-images.githubusercontent.com/67410847/194744339-3e010635-1edf-449f-89df-1b0bbce8a91e.png)

What goes wrong?
Thx for help
Mark

**Comment by benzino77** [comment]
Hi Mark,
If you want to use `development` branch of Tasmota you should rather use `development` branch of TasmoCompiler - ([description](https://github.com/benzino77/tasmocompiler#development-version)) which _should_ be compatible with Tasmota `development` branch (but this is development so you can expect unexpected from both sides: Tasmota and Tasmocompiler)
TasmoCompiler stable version is ment to be compatible with stable version of Tasmota (currently Tasmocompiler v9.1.0 is compatibile with Tasmota v12.x)

Couple of weeks ago there was a change introduced in Tasmota code: #299 regarding solo boards. This change is reflected on Tasmocompiler side in its development branch.

**Reply by Mark-Muc** [reply]
Hello benzino77
Many thanks for the quick answer and it worked without a problem!!
Regards 
Mark

**Reply by Mark-Muc** [reply]
Hello Benzino77!!
Today again i'd like to built a Solo1 Version .. but today I got the error:

```
Processing tasmota32solo1 (board: esp32_solo1_4M; framework: arduino; platform: https://github.com/tasmota/platform-espressif32/releases/download/v2.0.5.2/platform-espressif32-2.0.5.2.zip)
--------------------------------------------------------------------------------
Platform Manager: Installing https://github.com/tasmota/platform-espressif32/releases/download/v2.0.5.2/platform-espressif32-2.0.5.2.zip
Downloading 0% 10% 20% 30% 40% 50% 60% 70% 80% 90% 100%
Unpacking
Platform Manager: espressif32@2.0.5+2 has been installed!
Error: Unknown board ID 'esp32_solo1_4M'
Finished. Exit code: 1.
```
i have used the dev Version as u told me:
![image](https://user-images.githubusercontent.com/67410847/198871268-c143319b-97c9-487c-b427-6b3b7964351c.png)
https://3000-benzino77-tasmocompiler-rfjpy39411l.ws-eu73.gitpod.io/

Can u tell whats going wrong again??
Thank u very much!!
Mark-Muc

**Reply by benzino77** [reply]
It's pretty easy to explain. There are some changes on Tasmota side regarding platformio configuration. Those changes are not reflected on the TasmoCompiler side yet. For example the board name `esp32_solo1_4M` no longer exists in Tasmota platformio configuration - it is now called `esp32_solo1` (without `_4M` ate the end). That is why you are experiencing this error. BTW You are trying to use `development` branch so you should expect unexpected ;)

If you do not need changes introduced in Tasmota since version `v12.1.1` was released, you can go with Tasmota version `v12.1.1` to build your Solo1 board firmware. The version v12.2.0 will not work also, because there is a breaking change between v12.0.x/v12.1.x and v12.2.x - those versions are just incompatible from "compiling sources" perspective and TasmoCompiler cannot work differently based on version one choose. Generally when I drop support for particular version of Tasmota I change major version number of TasmoCompiler. So to be compatible with v12.2.x I should release TasmoCompiler version v10.0.0 and drop support for versions: v12.0.0, v12.0.1, v12.0.2, v12.1.0 and v12.1.1.

I started discussion some time ago [link](https://github.com/arendst/Tasmota/discussions/12516) to point out that introducing breaking changes without changing major version number is not the best practice ([link](https://semver.org/))

**Reply by Mark-Muc** [reply]
many thx for u r fast answer and it won't be any problem @ all .. just tell u ... .. i got a 12.1.xx version of the solo1 and that should do the work .. just was thinking to install the newest version . I am quite sure that it will work SOME day's!!
Again Thank u
Mark-Muc

**Reply by benzino77** [reply]
Development branch of TasmoCompiler is updated, so you can use docker image or gitpod to compile your binary.

**Comment by Jason2866** [comment]
A few words from the guy responsible for this ;-)
@benzino77  First, thank you for the great TasmoCompiler! It has closed the gap for users which needs a customized Tasmota version but do not want to dive in configuration and compiling. Just use Tasmota!

Thats the main goal of Tasmota. Install and use!

To reach this goal Theo and all people contributing to his project, try to include new devices / features AND
make the use of Tasmota even more user friendly. Sometimes breaking changes are needed to get there.

I know renaming and changing boards.json is a big change in the build process. But this change did not change the resulting firmwares. So why should the Tasmota version be changed?

Yes, it hits projects like yours to stop working.

With this change we reduced the needed boards.json and got more flexibilty to build variants more easy.
So there is now no more the need to know the real flash size of an ESP32x device. Tasmota detects and increases the Filesystem on first boot to thr max size. No harcoded webcam variant left. Now it is easy possible to compile a webcam build for different esp cams or esp32-S3 cams. 

To get there we had two possibilities add a boards.json for special esp32-x boards OR find a smart way to use a fixed amount of boards.json and add functions by configuration later (autoconfig).

This advantage is not fully visible now since this approach is at its beginning. First use cases are S3 cameras, and other devices based on the esp32-s3.
The S3 opened new possibilities. It is the first MCU which has enough resources overall to put a lot of features in the builded firmware and enable only the needed features (with the config at runtime).
So it is no problem to have the webcam driver in a built and flash this on a S3 which does not have a cam!
Other features and drivers are possible to load during runtime. Berry is the key, powerfull scripting language!
There is a for example a Berry PoC done for anouncing BLE devices to HomeAssistant (Autodiscovery)
Just a small BLE driver in Tasmota code needed.

Many words. In short, every change we made, we do with something in mind to make Tasmota better

**Reply by benzino77** [reply]
> A few words from the guy responsible for this ;-)

Jason! you don't have to feel sorry or responsible for something wrong (never ever) - you, and entire Tasmota Dev Team are doing _GREAT_  job. Frankly speaking _GREAT_ is not enough to describe what Tasmota is or what Tasmota become. It is absolutely fantastic.

> In short, every change we made, we do with something in mind to make Tasmota better

Tasmota is a "S" class IoT solution. Just _SUPERB_.  TasmoCompiler is "paying my debt" to community for using Tasmota. Keep doing what you are doing. I'll do my best, to keep TasmoCompiler usable and helpful (which means that I will complain sometimes here or there) ;)
